### PR TITLE
Added identiferEquals method for expressions to use fields with quoti…

### PR DESCRIPTION
**Pull request by skie**
…ng in join conditions. Added identifier parameter for expressions. Request created against master to replace #8238

### CODE REVIEW

File: src/Database/Expression/QueryExpression.php [hunk 1]
```diff
@@ -439,6 +439,18 @@ public function count()
...
     /**
+     * Builds equal condition or assignment with identifier wrapping.
+     *
+     * @param string $left Left join condition field name.
+     * @param string $right Right join condition field name.
+     * @return $this
+     */
+    public function equalFields($left, $right)
+    {
+        return $this->eq(new IdentifierExpression($left), new IdentifierExpression($right));
```
Comment (lorenzo) [inline src/Database/Expression/QueryExpression.php]: I think we we should test for $left and $right not being `ExpressionInterface` before wrapping them in an identifier expression

File: src/Database/Expression/CaseExpression.php [hunk 2]
```diff
@@ -130,6 +130,10 @@ protected function _addExpressions($conditions, $values, $types)
                 $value = $keyValues[$k];
                 array_push($this->_values, $value);
                 continue;
+            } elseif ($value === 'identifier') {
+                $value = new IdentifierExpression($keyValues[$k]);
+                array_push($this->_values, $value);
+                continue;
```
Comment (lorenzo) [inline src/Database/Expression/CaseExpression.php]: Can we add this case here too? https://github.com/cakephp/cakephp/blob/master/src/Database/Expression/FunctionExpression.php#L110

Comment (lorenzo) [inline src/Database/Expression/CaseExpression.php]: Never mind, you actualyl did it already :D

File: src/Database/Expression/CaseExpression.php [hunk 3]
```diff
@@ -130,6 +130,10 @@ protected function _addExpressions($conditions, $values, $types)
                 $value = $keyValues[$k];
                 array_push($this->_values, $value);
                 continue;
+            } elseif ($value === 'identifier') {
```
Comment (dereuromark) [inline src/Database/Expression/CaseExpression.php]: the else are not needed in that case, as it aborts early anyway.

Comment (skie) [inline src/Database/Expression/CaseExpression.php]: use of elseif is more compact. other reason i sued it there already in next lines
`
                continue;
            } elseif ($value instanceof ExpressionInterface) {
`

Comment (dereuromark) [inline src/Database/Expression/CaseExpression.php]: Yeah that's probably out of scope of this PR I think.
But in general else should be avoided wherever possible. Compactness is somewhat irrelevant here.

File: src/Database/Expression/QueryExpression.php [hunk 4]
```diff
@@ -439,6 +439,24 @@ public function count()
...
+     *
+     * @param string $left Left join condition field name.
+     * @param string $right Right join condition field name.
+     * @return $this
+     */
+    public function equalFields($left, $right)
+    {
+        $wrapIdentifier = function ($field) {
+            if ($field instanceof ExpressionInterface) {
+                return $field;
```
Comment (markstory) [inline src/Database/Expression/QueryExpression.php]: Do any of the tests get in here?

File: src/Database/Expression/CaseExpression.php (hunk 3 above)
Comment (lorenzo) [inline src/Database/Expression/CaseExpression.php]: In general we prefer another if clause

File: src/Database/Expression/QueryExpression.php (hunk 4 above)
Comment (skie) [inline src/Database/Expression/QueryExpression.php]: Yes, Query tests cover this method.

**Comment by markstory** [comment]
This looks good to me. Any thoughts @lorenzo ?

**Comment by lorenzo** [comment]
Woops, I thought it was merged already

**Pull request merged**